File: src/utils/searchContentUtils.ts
import { SearchContentResponse } from '../api/types/contentExploration.js';

export type ReducedSearchContentResponse = Partial<Record<SearchItemContent, unknown>>;

type SearchItemContent =
  | 'caption'
  | 'comments'
  | 'connectedWorkbooksCount'
  | 'connectionType'
  | 'containerName'
  | 'datasourceIsPublished'
  | 'datasourceLuid'
  | 'downstreamWorkbookCount'
  | 'extractCreationPending'
  | 'extractRefreshedAt'
  | 'extractUpdatedAt'
  | 'favoritesTotal'
  | 'hasActiveDataQualityWarning'
  | 'hasExtracts'
  | 'hasSevereDataQualityWarning'
  | 'hitsSmallSpanTotal'
  | 'hitsTotal'
  | 'isCertified'
  | 'isConnectable'
  | 'locationName'
  | 'luid'
  | 'modifiedTime'
  | 'ownerId'
  | 'ownerName'
  | 'parentWorkbookName'
  | 'projectId'
  | 'projectName'
  | 'sheetType'
  | 'tags'
  | 'title'
  | 'totalViewCount'
  | 'viewCountLastMonth'
  | 'type'
  | 'workbookDescription';

function getReducedSearchItemContent(
  content: Record<string, unknown>,
): ReducedSearchContentResponse {
  const r: ReducedSearchContentResponse = {};
  if (content.modifiedTime) r.modifiedTime = content.modifiedTime;
  if (content.sheetType) r.sheetType = content.sheetType;
  if (content.caption) r.caption = content.caption;
  if (content.workbookDescription) r.workbookDescription = content.workbookDescription;
  if (content.type) r.type = content.type;
  if (content.ownerId) r.ownerId = content.ownerId;
  if (content.title) r.title = content.title;
  if (content.ownerName) r.ownerName = content.ownerName;
  if (content.containerName) {
    if (content.type === 'view') {
      r.parentWorkbookName = content.containerName;
    } else {
      r.containerName = content.containerName;
    }
  }
  if (content.luid) r.luid = content.luid;
  if (content.locationName) r.locationName = content.locationName;
  if (Array.isArray(content.comments) && content.comments.length) r.comments = content.comments;
  if (content.hitsTotal != undefined) r.totalViewCount = content.hitsTotal;
  if (content.favoritesTotal != undefined) r.favoritesTotal = content.favoritesTotal;
  if (Array.isArray(content.tags) && content.tags.length) r.tags = content.tags;
  if (content.projectId) r.projectId = content.projectId;
  if (content.projectName) r.projectName = content.projectName;
  if (content.hitsSmallSpanTotal != undefined) r.viewCountLastMonth = content.hitsSmallSpanTotal;
  if (content.downstreamWorkbookCount != undefined) r.downstreamWorkbookCount = content.downstreamWorkbookCount;
  if (content.isConnectable != undefined) r.isConnectable = content.isConnectable;
  if (content.datasourceIsPublished != undefined) r.datasourceIsPublished = content.datasourceIsPublished;
  if (content.connectionType) r.connectionType = content.connectionType;
  if (content.isCertified != undefined) r.isCertified = content.isCertified;
  if (content.hasExtracts != undefined) r.hasExtracts = content.hasExtracts;
  if (content.extractRefreshedAt) r.extractRefreshedAt = content.extractRefreshedAt;
  if (content.extractUpdatedAt) r.extractUpdatedAt = content.extractUpdatedAt;
  if (content.connectedWorkbooksCount != undefined) r.connectedWorkbooksCount = content.connectedWorkbooksCount;
  if (content.extractCreationPending != undefined) r.extractCreationPending = content.extractCreationPending;
  if (content.hasSevereDataQualityWarning != undefined) r.hasSevereDataQualityWarning = content.hasSevereDataQualityWarning;
  if (content.datasourceLuid) r.datasourceLuid = content.datasourceLuid;
  if (content.hasActiveDataQualityWarning != undefined) r.hasActiveDataQualityWarning = content.hasActiveDataQualityWarning;
  return r;
}

export function reduceSearchContentResponse(
  response: SearchContentResponse,
): Array<ReducedSearchContentResponse> {
  let searchResults: Array<ReducedSearchContentResponse> = [];
  if (response.items) {
    for (const item of response.items) {
      searchResults.push(getReducedSearchItemContent(item.content as Record<string, unknown>));
    }
  }

  // Remove duplicate datasources with luid matching a unifieddatasource's datasourceLuid
  const unifiedDatasourceLuids = new Set(
    searchResults
      .filter((item) => item.type === 'unifieddatasource')
      .map((item) => item.datasourceLuid)
      .filter((luid): luid is string => typeof luid === 'string'),
  );

  searchResults = searchResults.filter((item) => {
    if (item.type === 'datasource') {
      return typeof item.luid === 'string' && !unifiedDatasourceLuids.has(item.luid);
    }
    return true;
  });

  // Normalize unifieddatasource entries to datasource entries
  for (const item of searchResults) {
    if (item.type === 'unifieddatasource') {
      item.type = 'datasource';
      item.luid = item.datasourceLuid;
      delete item.datasourceLuid;
    }
  }

  return searchResults;
}

File: src/tableau_cli/utils/search_content_utils.py
from __future__ import annotations

from typing import Any


def _get_reduced_search_item_content(content: dict[str, Any]) -> dict[str, Any]:
    r: dict[str, Any] = {}
    if content.get("modifiedTime"):
        r["modifiedTime"] = content["modifiedTime"]
    if content.get("sheetType"):
        r["sheetType"] = content["sheetType"]
    if content.get("caption"):
        r["caption"] = content["caption"]
    if content.get("workbookDescription"):
        r["workbookDescription"] = content["workbookDescription"]
    if content.get("type"):
        r["type"] = content["type"]
    if content.get("ownerId"):
        r["ownerId"] = content["ownerId"]
    if content.get("title"):
        r["title"] = content["title"]
    if content.get("ownerName"):
        r["ownerName"] = content["ownerName"]
    if content.get("containerName"):
        if content.get("type") == "view":
            r["parentWorkbookName"] = content["containerName"]
        else:
            r["containerName"] = content["containerName"]
    if content.get("luid"):
        r["luid"] = content["luid"]
    if content.get("locationName"):
        r["locationName"] = content["locationName"]
    if isinstance(content.get("comments"), list) and content["comments"]:
        r["comments"] = content["comments"]
    if content.get("hitsTotal") is not None:
        r["totalViewCount"] = content["hitsTotal"]
    if content.get("favoritesTotal") is not None:
        r["favoritesTotal"] = content["favoritesTotal"]
    if isinstance(content.get("tags"), list) and content["tags"]:
        r["tags"] = content["tags"]
    if content.get("projectId"):
        r["projectId"] = content["projectId"]
    if content.get("projectName"):
        r["projectName"] = content["projectName"]
    if content.get("hitsSmallSpanTotal") is not None:
        r["viewCountLastMonth"] = content["hitsSmallSpanTotal"]
    if content.get("downstreamWorkbookCount") is not None:
        r["downstreamWorkbookCount"] = content["downstreamWorkbookCount"]
    if content.get("isConnectable") is not None:
        r["isConnectable"] = content["isConnectable"]
    if content.get("datasourceIsPublished") is not None:
        r["datasourceIsPublished"] = content["datasourceIsPublished"]
    if content.get("connectionType"):
        r["connectionType"] = content["connectionType"]
    if content.get("isCertified") is not None:
        r["isCertified"] = content["isCertified"]
    if content.get("hasExtracts") is not None:
        r["hasExtracts"] = content["hasExtracts"]
    if content.get("extractRefreshedAt"):
        r["extractRefreshedAt"] = content["extractRefreshedAt"]
    if content.get("extractUpdatedAt"):
        r["extractUpdatedAt"] = content["extractUpdatedAt"]
    if content.get("connectedWorkbooksCount") is not None:
        r["connectedWorkbooksCount"] = content["connectedWorkbooksCount"]
    if content.get("extractCreationPending") is not None:
        r["extractCreationPending"] = content["extractCreationPending"]
    if content.get("hasSevereDataQualityWarning") is not None:
        r["hasSevereDataQualityWarning"] = content["hasSevereDataQualityWarning"]
    if content.get("datasourceLuid"):
        r["datasourceLuid"] = content["datasourceLuid"]
    if content.get("hasActiveDataQualityWarning") is not None:
        r["hasActiveDataQualityWarning"] = content["hasActiveDataQualityWarning"]
    return r


def reduce_search_content_response(response: dict[str, Any]) -> list[dict[str, Any]]:
    search_results: list[dict[str, Any]] = []
    items = response.get("items") or []
    for item in items:
        search_results.append(_get_reduced_search_item_content(item.get("content", {})))

    # Remove duplicate datasources with luid matching a unifieddatasource's datasourceLuid
    unified_datasource_luids: set[str] = set()
    for item in search_results:
        if item.get("type") == "unifieddatasource" and isinstance(item.get("datasourceLuid"), str):
            unified_datasource_luids.add(item["datasourceLuid"])

    search_results = [
        item
        for item in search_results
        if not (
            item.get("type") == "datasource"
            and isinstance(item.get("luid"), str)
            and item["luid"] in unified_datasource_luids
        )
    ]

    # Normalize unifieddatasource entries to datasource entries
    for item in search_results:
        if item.get("type") == "unifieddatasource":
            item["type"] = "datasource"
            item["luid"] = item.pop("datasourceLuid", None)

    return search_results
